rclcpp/include/rclcpp/intra_process_manager_impl.hpp

```diff
@@ -52,191 +51,189 @@ class IntraProcessManagerImplBase
-    if (size > std::numeric_limits<uint64_t>::max()) {
-      throw std::invalid_argument("the calculated buffer size is too large");
-    }
-    publishers_[id].sequence_number.store(0);
+    std::unordered_set<uint64_t> take_shared_subscriptions;
+    std::unordered_set<uint64_t> take_ownership_subscriptions;
+  };
 
-    publishers_[id].buffer = mrb;
-    publishers_[id].target_subscriptions_by_message_sequence.reserve(size);
-  }
+  template<typename T>
+  using RebindAlloc = typename std::allocator_traits<Allocator>::template rebind_alloc<T>;
 
-  void
-  remove_publisher(uint64_t intra_process_publisher_id)
+  using SubscriptionMap = std::unordered_map<
+    uint64_t, SubscriptionInfo,
+    std::hash<uint64_t>, std::equal_to<uint64_t>,
+    RebindAlloc<std::pair<const uint64_t, SubscriptionInfo>>>;
+
+  using PublisherMap = std::unordered_map<
+    uint64_t, PublisherInfo,
+    std::hash<uint64_t>, std::equal_to<uint64_t>,
+    RebindAlloc<std::pair<const uint64_t, PublisherInfo>>>;
+
+  using PublisherToSubscriptionIdsMap = std::unordered_map<
+    uint64_t, SplittedSubscriptions,
+    std::hash<uint64_t>, std::equal_to<uint64_t>,
+    RebindAlloc<std::pair<const uint64_t, SplittedSubscriptions>>>;
+
+  PublisherToSubscriptionIdsMap pub_to_subs_;
+  SubscriptionMap subscriptions_;
+  PublisherMap publishers_;
+
+  void insert_sub_id_for_pub(uint64_t sub_id, uint64_t pub_id, bool use_take_shared_method)
   {
-    publishers_.erase(intra_process_publisher_id);
+    if (use_take_shared_method) {
+      pub_to_subs_[pub_id].take_shared_subscriptions.insert(sub_id);
+    } else {
+      pub_to_subs_[pub_id].take_ownership_subscriptions.insert(sub_id);
+    }
   }
 
-  // return message_seq and mrb
-  mapped_ring_buffer::MappedRingBufferBase::SharedPtr
-  get_publisher_info_for_id(
-    uint64_t intra_process_publisher_id,
-    uint64_t & message_seq)
+  bool can_communicate(PublisherInfo pub_info, SubscriptionInfo sub_info)
   {
-    std::lock_guard<std::mutex> lock(runtime_mutex_);
-    auto it = publishers_.find(intra_process_publisher_id);
-    if (it == publishers_.end()) {
-      throw std::runtime_error("get_publisher_info_for_id called with invalid publisher id");
+    // publisher and subscription must be on the same topic
+    if (strcmp(pub_info.topic_name, sub_info.topic_name) != 0) {
+      return false;
+    }
+
+    // a reliable subscription can't be connected with a best effort publisher
+    if (sub_info.options.reliability == RMW_QOS_POLICY_RELIABILITY_RELIABLE &&
+      pub_info.options.reliability == RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT)
+    {
+      return false;
     }
-    PublisherInfo & info = it->second;
-    // Calculate the next message sequence number.
-    message_seq = info.sequence_number.fetch_add(1);
 
-    return info.buffer;
+    // a publisher and a subscription with different durability can't communicate
+    if (sub_info.options.durability != pub_info.options.durability) {
```

Comment:
I agree with you, but I haven't found such a function in the rmw.

I added a todo for now.
I was wondering if something like this should be implemented in the `rmw` or in the specific `rmw_implementations`...

Reply:
I think this one is a problem. It's something that was done inherently in the old intra process implementation. It should really be up to the rmw to decide this logic I think. I'll bring it up in the ROS 2 meeting.

Reply:
We decided to fix this after merging this, but we'll most likely add an rmw function for this and have a single implementation of it in `rmw` itself that we recommend all middleware implementations use, but it will still be possible for them to make their own if they wish (understanding that changing these assumptions may break tests and/or existing applications).